Mathlib/Tactic/Linter/DeprecateModule.lean

```diff
@@ -0,0 +1,138 @@
+/-
+Copyright (c) 2025 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import Mathlib.Init
+
+/-!
+#  The "deprecateModule" linter
+
+The "deprecateModule" linter emits a warning when a file that has been renamed or split
+is imported.
+
+The usage is as follows. Write
+```lean
+import B
+...
+import Z
+
+deprecate_module since yyyy-mm-dd
+```
+in module `A` with the expectation that `A` contains nothing else.
+This triggers the `deprecateModule` linter to notify every file with `import A`
+to instead import the *direct imports* of `A`, that is `B,...,Z`.
+-/
+
+open Lean Elab Command
+
+namespace Mathlib.Linter
+
+/--
+The `deprecateModule` linter emits a warning when a file that has been renamed or split
+is imported.
+-/
+register_option linter.deprecateModule : Bool := {
+  defValue := true
+  descr := "enable the deprecateModule linter"
+}
+
+/--
+Defines the `deprecateModuleExt` extension for adding a `HashSet` of pairs of
+* a module `Name` that has been deprecated and
+* an array of `Name`s of modules that should be imported instead
+
+to the environment.
+-/
+initialize deprecateModuleExt :
+    SimplePersistentEnvExtension (Name × Array Name) (Std.HashSet (Name × Array Name)) ←
+  registerSimplePersistentEnvExtension {
+    addImportedFn := fun as => as.foldl Std.HashSet.insertMany {}
+    addEntryFn := .insert
+  }
+
+/--
+`addModuleDeprecation` adds to the `deprecateModuleExt` extension the pair consisting of the
+current module name and the array of its direct imports.
+
+It ignores `Init`, since this is a special module that is expected to be imported by all files.
```

Comment:
Perhaps: "It ignores direct imports in `Init`"? (I was thinking whether you meant "the current module name in Init"... but that doesn't make much sense.)

Reply:
I wrote "It ignores the `Init` import", since usually `Init` is not explicitly written and it is implicitly imported anyway.